### ISSUE TITLE
Object casting

### ISSUE THREAD

**Issue by marcospassos** [belgattitude/soluble-japha#40]
We have an use case where we need to transmit a tree structure. Does the library support casting objects resulting from method calls to PHP plain classes (`stdClass`)?

If not, I would like to leave this feature proposal. Objects could be cast to PHP standard classes (plain objects) keeping only the public properties. It would make easy to transmit data between environments without implementing a custom serialization strategy.

**Comment by belgattitude** [comment]
Very good idea, I'll see in the comings days

**Comment by belgattitude** [comment]
Hi Marcos, 

I'm currently experimenting some kind of serialization with json-io and gson. 

Correct me if I'm wrong, but it looks to me that it would be the first step in order to get a valid representation of the Java object... Especially if you intend to work with deeply nested structures. 

The second step would be to get an stdClass representation of the Json (if still needed)... This part looks more easy to me.

Could you look at my latest commit for a proof of concept (need some material to test): https://github.com/belgattitude/soluble-japha/commit/e40cc6748969ed9fae77103b246ece93f9d22ff8
Especially this part: https://github.com/belgattitude/soluble-japha/blob/e40cc6748969ed9fae77103b246ece93f9d22ff8/test/src/SolubleTest/Japha/ThirdParty/JsonIoTest.php#L61 with json-io. I'll try to get a working example with gson soon.

I would really appreciate your feedback on this.

Thanks,

Seb

**Comment by marcospassos** [comment]
@belgattitude yes, we're doing the same here: we use Gson to serialize data on the Java side, and then deserialize the structure on the  PHP side. My point is that the library could provide a way to transmit data transparently. I'm not sure if the JSON.io is needed, perhaps only Java -> Json -> PHP (the type is not relevant, once it will become a plain class on the PHP side).

Still more interesting and fast would be serializing Java objects using the PHP's serialization format. I've found [a project for such purpose](https://github.com/kayahr/pherialize). Unfortunately, it looks abandoned but the code is still functional.

**Comment by belgattitude** [comment]
@marcospassos great... My personal feeling is that gson, jackson, json-io... will be far ahead that what I can develop in terms of performance and language features. Json or xml parser have fantastic parsers nowadays... Not even sure php serialization matches the performance.

Some time ago I was looking at protocol buffers, but cannot really use them in PHP without an extension (pure php is slow)... But just like you did, I would go the gson way (or json-io) and deserialize the json on the php side.

Will think about it. But what's sure is that we'll have to implement serialization on the java side (on the forked bridge https://github.com/belgattitude/php-java-bridge and eventually on the official version - looks alive now)...

**Comment by belgattitude** [comment]
@marcospassos FYI, I've added some docs for [object serialization](http://docs.soluble.io/soluble-japha/language_recipes/#json) and in the [optimization section](http://docs.soluble.io/soluble-japha/language_optimizations/#java-serialization) of the manual. That can be helpful for others.

Thanks for your remarks, it opens my mind ;)

**Comment by marcospassos** [comment]
Great work, @belgattitude! Thanks!

**Comment by belgattitude** [comment]
Hi marcos, I'll close for now... Feel free to reopen.

**Comment by marcospassos** [comment]
Hi @belgattitude!

Regarding the serialization approach that we briefly discussed, I've learned the hard way that serializing more than 50+ classes is a hit on performance for most of the serialization libraries written in PHP, including JMS Serializer and Symfony Serializer. Just to give you an idea, for deserializing 30+ distinct objects may take long seconds depending on the data complexity. Luckily, PHP's native serialization is pretty fast and the unserialize() function handles the same case in a few milliseconds.

Yesterday, I released the first version of the library that I'm using to encode Java object as PHP serialization format. Perhaps, mentioning it in the documentation may help other people who are facing the same issue.

For anyone interested, the library is available at https://github.com/marcospassos/java-php-serializer.
